### PR TITLE
GPX Support

### PR DESCRIPTION
**Pull request by dvdmrtnz**
I'm working on adding some basic GPX support to Organic Maps. This is still a work in progress...

https://github.com/organicmaps/organicmaps/issues/624

### CODE REVIEW

**Review by biodranik — COMMENTED** [review]
I would start with unit tests that can read a sample GPX file with _all_ features that OM supports at the moment. And then new features could be added to the sample file and the test could be expanded.
Then your work can be done in TDD manner:
- We help with tests review
- You fix failing tests by implementing missing functionality.

**Comment by dvdmrtnz** [comment]
> Cool! The idea is to read GPX and convert/store it in our internal KML-compatible format, right? And do the same for the export. Right?

Yes



> I would start with unit tests that can read a sample GPX file with _all_ features that OM supports at the moment. And then new features could be added to the sample file and the test could be expanded. Then your work can be done in TDD manner:
> 
> * We help with tests review
> * You fix failing tests by implementing missing functionality.

I added a very simple unit test as a starting point but I don't know how to run it. Could you guide me a little bit on how to run tests? I am currently working on a macOS computer with Xcode.

File: android/AndroidManifest.xml [hunk 1]
```diff
@@ -588,6 +588,18 @@
...
+      <intent-filter>
+        <action android:name="android.intent.action.VIEW" />
+        <category android:name="android.intent.category.DEFAULT" />
+        <category android:name="android.intent.category.BROWSABLE" />
+        <data android:scheme="content"/>
+        <data android:scheme="file"/>
+        <data android:scheme="data"/>
+        <data android:host="*"/>
+        <data android:mimeType="*/*"/>
+        <data android:pathPattern=".*\\.gpx" />
```
Comment (biodranik) [inline android/AndroidManifest.xml:600]: OSMand uses:

```
			<intent-filter
				android:label="@string/app_name"
				android:priority="50">
				<action android:name="android.intent.action.VIEW" />
				<category android:name="android.intent.category.DEFAULT" />
				<category android:name="android.intent.category.BROWSABLE" />
				<data android:scheme="file"/>
				<data android:scheme="content"/>
				<data android:host="*"/>
				<data android:mimeType="*/*"/>
				<data android:pathPattern=".*\\.gpx" />
				<data android:pathPattern=".*\\..*\\.gpx" />
				<data android:pathPattern=".*\\..*\\..*\\.gpx" />
				<data android:pathPattern=".*\\..*\\..*\\..*\\.gpx" />
				<data android:pathPattern=".*\\..*\\..*\\..*\\..*\\.gpx" />
				<data android:pathPattern=".*\\.gpx.zip" />
				<data android:pathPattern=".*\\..*\\.gpx.zip" />
				<data android:pathPattern=".*\\..*\\..*\\.gpx.zip" />
				<data android:pathPattern=".*\\..*\\..*\\..*\\.gpx.zip" />
				<data android:pathPattern=".*\\..*\\..*\\..*\\..*\\.gpx.zip" />
			</intent-filter>
```

File: map/bookmark_helpers.hpp [hunk 2]
```diff
@@ -68,11 +68,13 @@ enum class BookmarkBaseType : uint16_t
 extern std::string const kKmzExtension;
 extern std::string const kKmlExtension;
 extern std::string const kKmbExtension;
+extern std::string const kGpxExtension;
 extern std::string const kDefaultBookmarksFileName;
 
 enum class KmlFileType
 {
   Text,
+  Gpx,
```
Comment (biodranik) [inline map/bookmark_helpers.hpp:77]: Don't change the enum order, please. Add Gpx at the end.

File: map/bookmark_manager.cpp [hunk 3]
```diff
@@ -1869,7 +1869,17 @@ void BookmarkManager::LoadBookmarkRoutine(std::string const & filePath, bool isT
     std::string fileSavePath = GetKMLPath(filePath);
     if (!fileSavePath.empty())
     {
-      auto kmlData = LoadKmlFile(fileSavePath, KmlFileType::Text);
+      auto const ext = base::GetFileExtension(filePath);
+      std::unique_ptr<kml::FileData> kmlData;
+      if (ext == ".gpx")
+      {
```
Comment (biodranik) [inline map/bookmark_manager.cpp:1875]: Please check docs/CPP_STYLE.md and use it everywhere, to save the reviewer's and your time later.

**Review by biodranik — COMMENTED** [review]
You can select kml_tests as a target binary in XCode and run it.

To run tests from command line or Github Actions, please add your gpx_test.cpp into kml/kml_tests/CMakeLists.txt

**Comment by biodranik** [comment]
It is not ready and is not tested.

**Comment by AndrewShkrob** [comment]
From my pov, inserting `gpx` into `kml` types is really bad. We should consider separating them and providing a better architecture. 
Remember that you want to add GeoJSON format support as well.
What will be in the source code after it?
What will be in the source code after adding some more formats?
More `if`'s for checking for file type, more types in `KmlFileType`?
And all this logic in `bookmark_manager`?
That sucks

**Comment by biodranik** [comment]
The idea is to convert anything into the KML as a more broader format. And convert back into any format from KML.

**Comment by AndrewShkrob** [comment]
Right now there are no blockers. But think for the future. What if there will be such a format that cannot be converted to kml?
What will you do next? Add some more fields to the kml class to support converting with comments like `// @TODO fix in future`?
The problem with why we are still not supporting other file formats is bad architecture design.
Instead of working with some generic file interface, bookmark_manager works directly with the kml class.
And with this PR you won't fix the current problem. You'll make it even worse.

**Comment by dvdmrtnz** [comment]
> @dvdmrtnz @biodranik what is the current status of this PR? What is preventing us from merging?

@rtsisyk Currently, this PR is just a draft of the work necessary to implement GPX support in Organic Maps. 

I have not had time to work more on this PR so it is currently unfinished. If some other collaborator can continue working on GPX support they can use this PR as a base for their work if they think it is useful. Otherwise, I will work on this PR when I have some free time.

**Review by biodranik — COMMENTED** [review]
To anyone who wants to continue GPX implementation:

There is no need to store GPX file internally. KML file spec is broader than GPX, so let's keep using KML format under the hood, as it works now.

The idea to add GPX support in an easy way:
1. Implement the import _converter_ for GPX files: when any GPX file is opened, it should be converted to either internal structures and saved as KML (this approach will likely be faster), or converted directly to KML, and then "imported" as it was a KML file.
2. Implement unit tests to convert/parse GPX files.
3. (Can be done later, as a separate task) Implement GPX export in the same manner, by converting KML file into GPX format when a user pressed "export" button.

**Pull request closed without merging**